platform/api/esx/api.go

```diff
@@ -0,0 +1,759 @@
+	if err != nil {
+		return fmt.Errorf("uploading disks to ResourcePool: %v", err)
+	}
+	_ = object.NewVirtualMachine(a.client.Client, *entity)
+
+	return nil
+}
+
+func (a *API) TerminateDevice(vmName string) error {
+	finder, _, _, _, _, err := a.getServerDefaults()
+	if err != nil {
+		return fmt.Errorf("couldn't get server defaults: %v", err)
+	}
+
+	vm, err := finder.VirtualMachine(a.ctx, vmName)
+	if err != nil {
+		return fmt.Errorf("couldn't find VM: %v", err)
+	}
+
+	return a.deleteDevice(vm)
+}
+
+func (a *API) deleteDevice(vm *object.VirtualMachine) error {
+	task, err := vm.PowerOff(a.ctx)
+	if err != nil {
+		return fmt.Errorf("powering off vm: %v", err)
+	}
+
+	// We don't check for errors on this task because it will throw an error
+	// if the VM is already in a powered off state
+	_ = task.Wait(a.ctx)
+
+	task, err = vm.Destroy(a.ctx)
+	if err != nil {
+		return fmt.Errorf("destroying vm: %v", vm)
+	}
+
+	return task.Wait(a.ctx)
+}
+
+func (a *API) buildCISR(name string, flagOVAPath string, finder *find.Finder, defaultNetwork object.NetworkReference, resourcePool *object.ResourcePool, datastore *object.Datastore) (*Archive, *types.OvfCreateImportSpecResult, error) {
+	var nets []types.OvfNetworkMapping
+	nets = append(nets, types.OvfNetworkMapping{
+		Name:    "test",
+		Network: defaultNetwork.Reference(),
+	})
+
+	archive := &Archive{flagOVAPath}
+	envelope, err := archive.ReadEnvelope("*.ovf")
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading envelope: %v", err)
+	}
+
+	ovfHandler := object.NewOvfManager(a.client.Client)
+	cisp := types.OvfCreateImportSpecParams{
+		EntityName:         name,
+		OvfManagerCommonParams: types.OvfManagerCommonParams{
+			Locale:           "US"},
+		PropertyMapping: []types.KeyValue{},
+		NetworkMapping:  networkMap(finder, envelope),
+	}
+
+	descriptor, err := archive.ReadOvf("*.ovf")
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading ovf: %v", err)
+	}
+	cisr, err := ovfHandler.CreateImportSpec(a.ctx, string(descriptor), resourcePool, datastore, cisp)
+	if err != nil {
+		return nil, nil, err
+	}
+	if cisr.Error != nil {
+		return nil, nil, errors.New(cisr.Error[0].LocalizedMessage)
+	}
+	return archive, cisr, nil
+}
+
+func (a *API) getServerDefaults() (*find.Finder, *object.Datacenter, *object.ResourcePool, *object.Datastore, object.NetworkReference, error) {
+	finder := find.NewFinder(a.client.Client, true)
+	datacenter, err := finder.DefaultDatacenter(a.ctx)
+	if err != nil {
+		return nil, nil, nil, nil, nil, err
+	}
+	finder.SetDatacenter(datacenter)
+	resourcePool, err := finder.DefaultResourcePool(a.ctx)
+	if err != nil {
+		return nil, nil, nil, nil, nil, err
+	}
+	datastore, err := finder.DefaultDatastore(a.ctx)
+	if err != nil {
+		return nil, nil, nil, nil, nil, err
+	}
+
+	defaultNetwork, err := finder.DefaultNetwork(a.ctx)
+	if err != nil {
+		return nil, nil, nil, nil, nil, err
+	}
+
+	return finder, datacenter, resourcePool, datastore, defaultNetwork, nil
```

Comment:
Perhaps a `serverDefaults` struct would be a saner way to hold on to all of these values?